06_gpu_and_ml/yolo/yolo.py

```diff
@@ -0,0 +1,288 @@
+    model = YOLO('yolov10l.pt') # This downloads the model's base weights if not already present in the image.
+    return model
+
+yolo_image = (
+    modal.Image.from_registry("pytorch/pytorch:2.3.1-cuda11.8-cudnn8-runtime")
+    .apt_install(["libgl1-mesa-glx", "libglib2.0-0"])
+    .pip_install(["ultralytics", "datasets", "opencv-python"])
+    .run_function(load_yolo_weights)
+
+app = modal.App("train-yolo")
+
+# ## Downloading The Dataset
+# 
+# Similar to downloading model weights at build-time, it may be tempting to also download the DoclayNet dataset at build-time.
+# We instead choose to download the dataset into a Modal Volume, using a Modal Function.
+# This approach has two benefits:
+# - Reduces the size of the container image, making cold boots much faster. This is especially important for inference when there's no need for having the data in the container.
+# - Makes it easier to train on many datasets, storing to create many finetunes usable for inference.
+
+# Define a Volume, which will persist training data and trained weights between invocations 
+volume = modal.Volume.from_name("doclaynet-base-yolo", create_if_missing=True)
+
+def download_file(url, output_path):
+    from tqdm import tqdm
+    class DownloadProgressBar(tqdm):
+        def update_to(self, b=1, bsize=1, tsize=None):
+            if tsize is not None:
+                self.total = tsize
+            self.update(b * bsize - self.n)
+    with DownloadProgressBar(unit='B', unit_scale=True, miniters=1, desc=url.split('/')[-1]) as t:
+        urllib.request.urlretrieve(url, filename=output_path, reporthook=t.update_to)
+
+def convert_dataset_to_yolo(dataset_path, output_path):
+    from tqdm import tqdm
+    root_folder = Path(output_path)
+    if not root_folder.exists():
+        root_folder.mkdir(parents=True)
+
+    category_map = {
+                    "0": "Caption", "1": "Footnote", "2": "Formula", "3": "List-item",
+                    "4": "Page-footer", "5": "Page-header", "6": "Picture",
+                    "7": "Section-header", "8": "Table", "9": "Text", "10": "Title",
+                }
+    inverse_category_map = {v: k for k, v in category_map.items()}
+    with open(root_folder / "data.yaml", "w") as f:
+        yaml.dump(
+            {
+                "path": root_folder.absolute().as_posix(),
+                "train": "./images/train",
+                "val": "./images/val",
+                "test": "./images/test",
+                "names": category_map,
+            },
+            f,
+        )
+ 
+    for split in ["train", "val", "test"]:
+        print(f"Converting {split} dataset...")
+        
+        ann_dir = os.path.join(dataset_path, f"base_dataset/{split}/annotations")
+        img_dir = os.path.join(dataset_path, f"base_dataset/{split}/images")
+        
+        os.makedirs(root_folder / "labels" / split, exist_ok=True)
+        os.makedirs(root_folder / "images" / split, exist_ok=True)
```

Comment:
Slight preference for using `Path`-baed operations everywhere instead of `os` etc.